my/core/structure.py

```diff
@@ -0,0 +1,150 @@
+        └── profile
+            └── settings.json
+
+    Giving the top directory as the base, and some expected relative path like:
+
+    with match_structure(Path("export_dir"), expected=("messages/messages.csv", "index.json")) as results:
+        # results in this block is (Path("export_dir/exp_2020"), Path("export_dir/exp_2021"))
+
+    This doesn't require an exhaustive list of expected values, but its a good idea to supply
+    a complete picture of the expected structure to avoid false-positives
+
+    This does not recursively unzip zipfiles in the subdirectories,
+    it only unzips into a temporary directory if 'base' is a zipfile
+
+    A common pattern for using this might be to use get_files to get a list
+    of zipfiles or top-level gdpr export directories, and use match_structure
+    to search the resulting paths for a export structure you're expecting
+    """
+    from . import core_config as CC
+
+    tdir = CC.config.get_tmp_dir()
+
+    if isinstance(expected, str):
+        expected = (expected,)
+
+    is_zip: bool = base.suffix in ZIP_EXT
+
+    searchdir: Path = base.absolute()
+    try:
+        # if the file given by the user is a zipfile, create a temporary
+        # directory and extract the zipfile to that temporary directory
+        #
+        # this temporary directory is removed in the finally block
+        if is_zip:
+            # sanity check before we start creating directories/rm-tree'ing things
+            assert base.exists(), f"zipfile at {base} doesn't exist"
+
+            searchdir = Path(tempfile.mkdtemp(dir=tdir))
+
+            zf = zipfile.ZipFile(base)
+            zf.extractall(path=str(searchdir))
+
+        else:
+            if not searchdir.is_dir():
+                raise NotADirectoryError(f"Expected either a zipfile or a directory, received {searchdir}")
+
+        matches: List[Path] = []
+        possible_targets: List[Path] = [searchdir]
+
+        while len(possible_targets) > 0:
+            p = possible_targets.pop(0)
+
+            # factored out into a function to avoid weird stuff with continues/loop state
+            if _structure_exists(p, expected):
+                matches.append(p)
+            else:
+                # extend the list of possible targets with any subdirectories
+                for f in os.scandir(p):
+                    if f.is_dir():
+                        possible_targets.append(p / f.name)
+
+        if len(matches) == 0:
+            core_warnings.high(f"""While searching {base}, could not find a matching folder structure. Expected {expected}. You're probably missing required files in the gdpr/export""")
+
+        yield tuple(matches)
+
+    finally:
+
+        if is_zip:
+            # make sure we're not mistakenly deleting data
+            assert str(searchdir).startswith(str(tdir)), f"Expected the temporary directory for extracting zip to start with the temporary directory prefix ({tdir}), found {searchdir}"
```

Comment:
The context manager looks good! It also seems safe because of this assert; even though in theory it's doesn't look really necessary in the current version of code, if someone added some code between line 113-115, and it threw exception, it would be possible to delete the actual zip file by accident.
Such condition cleanup arises now and then and I prefer to handle it slightly differently


Although I feel like for such cases it's cleaner to rely on context managers as well. So I'd rewrite it as something like: 

```
@contextmanager
def ctx_zip():
    with ZipFile(base.absolute()) as zf, tempfile.TemporaryDirectory(dir=tdir) as td:
        # technically, ZipFile ctx manager is not necessarily since we're reading only?
        # but still, I guess it's nice to call close()
        zf.extractall(path=str(td))
        yield Path(td)

@contextmanager
def ctx_dir():
    yield base.absolute()

ctx = ctx_zip if is_zip else ctx_dir

with ctx as searchdir:
    # rest of code in try section (lines 124-142)

```

That way all the logic for each case is local and it's easier to reason about it

Reply:
Ah right, I knew there was something I could do with yielding back the location to make it generic across dir/zipfile -- agreed it is cleaner, not sure if thats more or less understandable for someone unfamiliar with contextmanagers trying to read this code though

I'm fine leaving it as is if you are

Reply:
yeah sure!